index.d.ts

```diff
@@ -1,10 +1,10 @@
-type JestResolverOptions = {
+type JestResolverOptions<T> = {
   basedir: string;
-  defaultResolver: (request: string, opts: any) => string,
+  defaultResolver: T;
   extensions?: Array<string>,
 };
 
-export default function resolve(
-  request: string,
-  options: JestResolverOptions,
-): string;
+export default function resolve<D extends (request: string, options: JestResolverOptions<D>) => any>(
```

Comment:
Is `any` a good default?  Should it be a string?

Reply:
That was the original rub, that the `defaultResolverAsync` returned a promise and not a string.  Here's a playground I threw together, and I'm not sure how to specify anything other than `any` here.  (I tried `unknown` and `string | Promise<string>`)

https://tsplay.dev/N7V04w

Reply:
A resolver should always return a `string` or `null` (from memory, I can very tomorrow)

Reply:
Even an async resolver?  Wouldn't it return a promise?

Reply:
Perhaps something like this: https://www.typescriptlang.org/play?target=99#code/C4TwDgpgBAShDOB7ANgNwgJwDwBUB8UAvFABQYQCOArgsAFxTzAYCWAdgOYA0UiYwLRG3gMAUrThI0mAPL9Bw3HgCURAjgDcAKC0BjIUygBDeCDa7JKdBgaXp2AAoZEAWxbwIWJq054CRUlVCAidXdwgAOnIpdBIAcgAzREQ45W19YWBGMwsEK0xbPPsvZnYOAmISIIJE5LjtLVBIKHEmO2s5AQMlAIBvLSgoABMIBKMqZGB2gtgi6yVtAF8GhKpzLrYoaPylMkoaJgZvMp4+DZEWiTnZeW78ZQYcKH7B8mAqDE2zhXgIkbGJlNrhg9tRaKdbsI0lpFjoMoYTDlJICAttpPE4jxev9xpNpjZjKZzPjFsoBoMKYMAPRUgB6AH44QYskTcvAUcQ0bE4pjnjjAfijkjgaTyZTqXTGVogA